commit ebeb0bbc7ae64007ba65cedd2542876a328eb9fd
Author: Michał Dulko <michal.dulko@intel.com>
Date:   Fri Sep 2 11:54:20 2016 +0200

    Fix CapacityWeiger to accept None volume type
    
    When fixing CapacityFilter in Icfc6ba5f47c3a61b49cbec5110ef7df05eb47195
    I haven't noticed that CapacityWeigher is subject to the same issue.
    This commit fixes that and (finally):
    
    Closes-Bug: 1619246
    Change-Id: I4b52aad87116472872efec4cefb2ac330c47c85b

diff --git a/cinder/scheduler/weights/capacity.py b/cinder/scheduler/weights/capacity.py
index 82f30ce..61da824 100644
--- a/cinder/scheduler/weights/capacity.py
+++ b/cinder/scheduler/weights/capacity.py
@@ -119,7 +119,7 @@ class CapacityWeigher(weights.BaseHostWeigher):
             # provisioned_capacity_gb to determine whether a volume can be
             # provisioned. Instead free capacity will be used to evaluate.
             thin = True
-            vol_type = weight_properties.get('volume_type', {})
+            vol_type = weight_properties.get('volume_type', {}) or {}
             provision_type = vol_type.get('extra_specs', {}).get(
                 'provisioning:type')
             if provision_type == 'thick':
diff --git a/cinder/tests/unit/scheduler/test_capacity_weigher.py b/cinder/tests/unit/scheduler/test_capacity_weigher.py
index 1c5dff6..b15823d 100644
--- a/cinder/tests/unit/scheduler/test_capacity_weigher.py
+++ b/cinder/tests/unit/scheduler/test_capacity_weigher.py
@@ -64,25 +64,19 @@ class CapacityWeigherTestCase(test.TestCase):
     # free = free_space - math.floor(total * reserved)
 
     @ddt.data(
-        {'type_key': 'provisioning:type', 'type_val': 'thin',
-         'vol_type': 'volume_type', 'extra_specs': 'extra_specs',
+        {'volume_type': {'extra_specs': {'provisioning:type': 'thin'}},
          'winner': 'host2'},
-        {'type_key': 'provisioning:type', 'type_val': 'thick',
-         'vol_type': 'volume_type', 'extra_specs': 'extra_specs',
+        {'volume_type': {'extra_specs': {'provisioning:type': 'thick'}},
          'winner': 'host1'},
-        {'type_key': None, 'type_val': None,
-         'vol_type': 'volume_type', 'extra_specs': 'extra_specs',
+        {'volume_type': {'extra_specs': {}},
          'winner': 'host2'},
-        {'type_key': None, 'type_val': None,
-         'vol_type': 'volume_type', 'extra_specs': None,
+        {'volume_type': {},
          'winner': 'host2'},
-        {'type_key': None, 'type_val': None,
-         'vol_type': None, 'extra_specs': None,
+        {'volume_type': None,
          'winner': 'host2'},
     )
     @ddt.unpack
-    def test_default_of_spreading_first(self, type_key, type_val,
-                                        vol_type, extra_specs, winner):
+    def test_default_of_spreading_first(self, volume_type, winner):
         hostinfo_list = self._get_all_hosts()
 
         # Results for the 1st test
@@ -108,11 +102,7 @@ class CapacityWeigherTestCase(test.TestCase):
         # so, host2 should win:
         weight_properties = {
             'size': 1,
-            vol_type: {
-                extra_specs: {
-                    type_key: type_val,
-                }
-            }
+            'volume_type': volume_type,
         }
         weighed_host = self._get_weighed_hosts(
             hostinfo_list,
@@ -121,25 +111,19 @@ class CapacityWeigherTestCase(test.TestCase):
         self.assertEqual(winner, utils.extract_host(weighed_host.obj.host))
 
     @ddt.data(
-        {'type_key': 'provisioning:type', 'type_val': 'thin',
-         'vol_type': 'volume_type', 'extra_specs': 'extra_specs',
+        {'volume_type': {'extra_specs': {'provisioning:type': 'thin'}},
          'winner': 'host4'},
-        {'type_key': 'provisioning:type', 'type_val': 'thick',
-         'vol_type': 'volume_type', 'extra_specs': 'extra_specs',
+        {'volume_type': {'extra_specs': {'provisioning:type': 'thick'}},
          'winner': 'host2'},
-        {'type_key': None, 'type_val': None,
-         'vol_type': 'volume_type', 'extra_specs': 'extra_specs',
+        {'volume_type': {'extra_specs': {}},
          'winner': 'host4'},
-        {'type_key': None, 'type_val': None,
-         'vol_type': 'volume_type', 'extra_specs': None,
+        {'volume_type': {},
          'winner': 'host4'},
-        {'type_key': None, 'type_val': None,
-         'vol_type': None, 'extra_specs': None,
+        {'volume_type': None,
          'winner': 'host4'},
     )
     @ddt.unpack
-    def test_capacity_weight_multiplier1(self, type_key, type_val,
-                                         vol_type, extra_specs, winner):
+    def test_capacity_weight_multiplier1(self, volume_type, winner):
         self.flags(capacity_weight_multiplier=-1.0)
         hostinfo_list = self._get_all_hosts()
 
@@ -166,11 +150,7 @@ class CapacityWeigherTestCase(test.TestCase):
         # so, host4 should win:
         weight_properties = {
             'size': 1,
-            vol_type: {
-                extra_specs: {
-                    type_key: type_val,
-                }
-            }
+            'volume_type': volume_type,
         }
         weighed_host = self._get_weighed_hosts(
             hostinfo_list,
@@ -179,25 +159,19 @@ class CapacityWeigherTestCase(test.TestCase):
         self.assertEqual(winner, utils.extract_host(weighed_host.obj.host))
 
     @ddt.data(
-        {'type_key': 'provisioning:type', 'type_val': 'thin',
-         'vol_type': 'volume_type', 'extra_specs': 'extra_specs',
+        {'volume_type': {'extra_specs': {'provisioning:type': 'thin'}},
          'winner': 'host2'},
-        {'type_key': 'provisioning:type', 'type_val': 'thick',
-         'vol_type': 'volume_type', 'extra_specs': 'extra_specs',
+        {'volume_type': {'extra_specs': {'provisioning:type': 'thick'}},
          'winner': 'host1'},
-        {'type_key': None, 'type_val': None,
-         'vol_type': 'volume_type', 'extra_specs': 'extra_specs',
+        {'volume_type': {'extra_specs': {}},
          'winner': 'host2'},
-        {'type_key': None, 'type_val': None,
-         'vol_type': 'volume_type', 'extra_specs': None,
+        {'volume_type': {},
          'winner': 'host2'},
-        {'type_key': None, 'type_val': None,
-         'vol_type': None, 'extra_specs': None,
+        {'volume_type': None,
          'winner': 'host2'},
     )
     @ddt.unpack
-    def test_capacity_weight_multiplier2(self, type_key, type_val,
-                                         vol_type, extra_specs, winner):
+    def test_capacity_weight_multiplier2(self, volume_type, winner):
         self.flags(capacity_weight_multiplier=2.0)
         hostinfo_list = self._get_all_hosts()
 
@@ -224,11 +198,7 @@ class CapacityWeigherTestCase(test.TestCase):
         # so, host2 should win:
         weight_properties = {
             'size': 1,
-            vol_type: {
-                extra_specs: {
-                    type_key: type_val,
-                }
-            }
+            'volume_type': volume_type,
         }
         weighed_host = self._get_weighed_hosts(
             hostinfo_list,
